feat(matchmake-extension): Implement GetPlayingSession

DaniElectra opened the pull request:
<!--

* Before making a pull request, ensure the changes are for an approved issue.
* If your changes are not for an approved issue, your pull request can and will be rejected.
*
* CHECK https://github.com/PretendoNetwork/REPO_NAME/issues?q=is%3Aopen+is%3Aissue+label%3Aapproved
* FOR APPROVED ISSUES!

-->

Resolves #XXX

### Changes:

<!--

* Describe your changes in as much detail as possible. Make sure to list your changes, as well as the rationale behind them.
* If applicable, include code snippets, images, videos, etc.
*
* If your changes require any database migrations, describe them in detail and leave any migration queries inside of a code
* block within a <details> tag.

-->

Reuse the code from the `GetSimplePlayingSession` method, assuming it behaves in a similar way. Also make the cleanup functions optional, as they are only intended for special cases like region locking in a game or similar.

Haven't tested the changes tbh, but I trust myself that this should be a simple change and work as expected.

- [x] I have read and agreed to the [Code of Conduct](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CODE_OF_CONDUCT.md).
- [x] I have read and complied with the [contributing guidelines](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CONTRIBUTING.md).
- [ ] What I'm implementing was an [approved issue](../issues?q=is%3Aopen+is%3Aissue+label%3Aapproved).
- [ ] I have tested all of my changes.

DaniElectra committed:
feat(matchmake-extension): Implement GetPlayingSession

Reuse the code from the `GetSimplePlayingSession` method, assuming it
behaves in a similar way.

DaniElectra committed:
chore(matchmake-extension): Make cleanup functions optional

The cleanup functions are only supposed to be used on special cases such
as removing region locking and similar.

TraceEntertains commented:
IIRC this has been tested in a couple games now, and seems to work fine, there is a strange bug in one game tested, but it is unlikely to have anything to do with this method lol (game doesn't show gatherings in a list even when server returns one), but in the same game with different MatchmakeSession info it works perfectly fine, output has been confirmed to be 1 to 1 for the original (minus structure versions)

jonbarrow commented on matchmake-extension/get_playing_session.go:
```diff
@@ -0,0 +1,51 @@
+
+	commonProtocol.manager.Mutex.RLock()
+
+	playingSessions, nexError := database.GetPlayingSession(commonProtocol.manager, lstPID)
```
`lstPID` is missing a length check. Only up to [`MAX_PRINCIPALID_SIZE_TO_FIND_MATCHMAKE_SESSION`](https://github.com/tech-ticks/RTDXTools/blob/232a7797e01369e9c12704f58fdde65dd3ac1c32/Assets/Scripts/Stubs/Generated/Assembly-CSharp/NexPlugin/Matchmake.cs#L30) can be requested at a time. I forgot which error is returned but `InvalidArgument` should be fine

jonbarrow commented on matchmake-extension/database/get_playing_session.go:
```diff
@@ -0,0 +1,93 @@
+
+// GetPlayingSession returns the playing sessions of the given PIDs
+func GetPlayingSession(manager *common_globals.MatchmakingManager, listPID []types.PID) ([]match_making_types.PlayingSession, *nex.Error) {
+	playingSessions := make([]match_making_types.PlayingSession, 0)
```
In my testing only up to [`MAX_MATCHMAKE_SESSION_BY_PARTICIPANT`](https://github.com/tech-ticks/RTDXTools/blob/232a7797e01369e9c12704f58fdde65dd3ac1c32/Assets/Scripts/Stubs/Generated/Assembly-CSharp/NexPlugin/Matchmake.cs#L32) is returned

jonbarrow commented on matchmake-extension/database/get_playing_session.go:
```diff
@@ -0,0 +1,93 @@
+		resultMatchmakeSession := match_making_types.NewMatchmakeSession()
+		var resultMatchmakeParam []byte
+
+		err := manager.Database.QueryRow(`SELECT
+			g.id,
+			g.owner_pid,
+			g.host_pid,
+			g.min_participants,
+			g.max_participants,
+			g.participation_policy,
+			g.policy_argument,
+			g.flags,
+			g.state,
+			g.description,
+			array_length(g.participants, 1),
+			g.started_time,
+			ms.game_mode,
+			ms.attribs,
+			ms.open_participation,
+			ms.matchmake_system_type,
+			ms.application_buffer,
+			ms.progress_score,
+			ms.option_zero,
+			ms.matchmake_param,
+			ms.refer_gid,
+			ms.user_password_enabled,
+			ms.system_password_enabled,
+			ms.codeword
+		FROM matchmaking.gatherings AS g
+		INNER JOIN matchmaking.matchmake_sessions AS ms ON ms.id = g.id
+		WHERE
+		g.registered=true AND
+		g.type='MatchmakeSession' AND
+		$1=ANY(g.participants)`, pid).Scan(
```
You can get the session info for anyone, but if you request info for a user who has you blocked then that users info is not returned. At the very least we should make a note of this somewhere, if not adding it now

jonbarrow commented on matchmake-extension/get_playing_session.go:
```diff
@@ -0,0 +1,51 @@
+
+	commonProtocol.manager.Mutex.RUnlock()
+
+	lstPlayingSession := types.NewList[match_making_types.PlayingSession]()
+	lstPlayingSession = playingSessions
```
You can avoid this by just returning the `types.List` from the database method. The `DataStore` rework PR does this in a few places. `types.List` can be used as a normal slice/array type

jonbarrow reviewed (changes requested):
Looks good! Just missing a few sanity checks/notes

DaniElectra committed:
fix(matchmake-extension): Add sanity checks to GetPlayingSession

Given the information provided it's possible for participants to be in
multiple matchmake sessions, so reorganize the database code to handle
that.

jonbarrow commented:
LGTM! I'll approve this now but won't merge yet until given the greenlight, incase someone wants to test the latest changes

jonbarrow commented on matchmake-extension/database/get_playing_session.go:
```diff
@@ -0,0 +1,108 @@
+
+// GetPlayingSession returns the playing sessions of the given PIDs
+func GetPlayingSession(manager *common_globals.MatchmakingManager, listPID types.List[types.PID]) (types.List[match_making_types.PlayingSession], *nex.Error) {
+	playingSessions := make([]match_making_types.PlayingSession, 0)
```
I didn't catch this actually, this can be improved. In Go if a slice is made without any capacity set then whenever you `append` to the slice it will double the capacity, reallocate the memory, and then copy the contents of the old slice to the new one. Since we know that only 1000 entries can exist here, we can go ahead and pre-allocate the slice to avoid this overhead. Preallocating the slice *could* waste some memory for slices smaller than 1000, but it avoids the need for multiple reallocations and copies

DaniElectra replied:
Ah right, I didn't add it at first since it would be wasting memory, but I'll go ahead and add it

DaniElectra committed:
fix(matchmake-extension): Allocate max capacity for GetPlayingSession

TraceEntertains commented:
Tested on Yo-kai Watch Blasters (before slice pre-allocation change, but I don't think that would break it), seems to work fine
